File: hcemi-appdev-pims/MI.PIMS.UI/MI.PIMS.UI/wwwroot/js/Telemetry/TelemetryLogger.js
class TelemetryLogger {
    constructor() {
        this.appInsights = window.appInsights;
        this.isEnabled = false;
        if (window.insightsEnabled !== undefined && window.insightsEnabled !== "") {
            this.isEnabled = window.insightsEnabled;
        }
    };

    trackEvent(eventName, logProperties) {
        if (this.appInsights !== undefined) {
            const properties = logProperties.toTelemetryProperties();
            this.appInsights.trackEvent({
                name: eventName,
                properties: properties
            });
        }
    };

    trackTrace(message, severity, logProperties) {
        if (this.appInsights !== undefined) {
            const properties = logProperties.toTelemetryProperties();
            this.appInsights.trackTrace({
                message: message,
                severity: severity,
                properties: properties
            });
        }
    };

    trackException(exception, logProperties) {
        if (this.appInsights !== undefined) {
            const properties = logProperties.toTelemetryProperties();
            this.appInsights.trackException({
                exception: exception,
                properties: properties
            });
        }
    };

    trackMetric(name, value) {
        if (this.appInsights !== undefined) {
            this.appInsights.trackMetric({
                name: name,
                average: value
            });
        }
    };

    isAppInsightsEnabled(key) {
        if (this.isEnabled && key) {
            return true;
        }
        return false;
    }
};

File: hcemi-appdev-pims/MI.PIMS.MT/MI.PIMS.BL/Common/Telemetry/TelemetryLogger.cs
namespace MI.PIMS.BL.Common.Telemetry
{
    using Microsoft.ApplicationInsights;
    using Microsoft.ApplicationInsights.DataContracts;
    using Microsoft.Extensions.Logging;
    using System;
    using System.Diagnostics;

    public class TelemetryLogger<T>
    {
        private readonly TelemetryClient _telemetryClient;
        private readonly ILogger<T> _logger;
        private readonly Helper _helper;
        private string _MSID = string.Empty;

        public TelemetryLogger(TelemetryClient telemetryClient, ILogger<T> logger, Helper helper)
        {
            _telemetryClient = telemetryClient;
            _logger = logger;
            _helper = helper;

            try
            {
                _MSID = _helper.GetMSID();
            }
            catch (Exception) {
                _MSID = "";
            }

            _telemetryClient.Context.User.AuthenticatedUserId = _MSID;
        }

        public string GetAppInsightsOperationId
        {
            get
            {
                try
                {
                    var activity = Activity.Current;
                    return activity?.RootId ?? activity?.Id;
                }
                catch
                {
                    return string.Empty;
                }
            }
        }

        public void TrackEvent(string eventName, LogInfo log = null)
        {
            _logger.LogInformation("Telemetry Event: {EventName}", eventName);
            _telemetryClient.TrackEvent(eventName, log.ToDictionary());
        }

        public void TrackTrace(string message, SeverityLevel severity = SeverityLevel.Information, LogInfo log = null)
        {
            _logger.LogInformation("Telemetry Trace: {Message}", message);
            _telemetryClient.TrackTrace(message, severity, log.ToDictionary());
        }

        public void TrackException(Exception ex, LogInfo log = null)
        {
            _logger.LogError(ex, "Telemetry Exception: {Message}", ex.Message);
            _telemetryClient.TrackException(ex, log.ToDictionary());
        }

        public void TrackMetric(string name, double value)
        {
            _logger.LogInformation("Telemetry Metric: {Metric} = {Value}", name, value);
            _telemetryClient.GetMetric(name).TrackValue(value);
        }
    }
}